### ISSUE TITLE
Download.php command injection

### ISSUE THREAD

**Issue by wireghoul** [Codiad/Codiad#808]
Hi there,

It appears your download script suffers from command injection, to illustrate the issue create a directory named ";uname -a" and try to download it.

Cheers,
Wireghoul

**Comment by evertton** [comment]
The problem is here: [filemanager/download.php#L50](https://github.com/Codiad/Codiad/blob/master/components/filemanager/download.php#L50)
Sanitize the files and directories names should be sufficient. We can sanitize using this regular expression `[a-zA-Z0-9\s_-]+`.

**Comment by daeks** [comment]
Yeah, I have already seen it, but marked as minor cause it requires some prerequirements which are not critical. There is already a function like that available in the controller, so I would like to use that one. 

Anyway, please dont post security issues or examples to public github issues :) - pull in a PR or by email ;)

**Comment by wireghoul** [comment]
I would, but your website doesn't list any email contact at all. The issue requires an authenticated session and most importantly searching github for the words "system" and  "$_GET" will serve up the vulnerable code. Combined with the number of bots that scrape github for this type of coding practices it seemed reasonable enough to post it to your issue tracker.

**Comment by daeks** [comment]
Posting that issue was quite good, just removed the direct link from evertton :)
I had a look at the webpage itself, you are right. it is not mentioned there (only on github https://github.com/Codiad/ and on facebook)

**Comment by evertton** [comment]
Oh, sorry! I completely forgot that we should not show bugs before correction. :dizzy_face: 
Please, check my PR. :smile:

**Comment by wireghoul** [comment]
Unfortunately the fix is flawed, escapeshell args does not cover argument injection and specifying a command to run with the -I argument ($_GET['path']='-I /bin/evil' dir/) will let an attacker divert the tar process to execute a command of their choosing. See the following article for more details: http://www.cso.com.au/vendor_blog/14/bae-systems-detica/5342/security-issues-with-using-phps-escapeshellarg/
The whitelisting sanitation approach initially suggested was a much safer solution to this problem.

**Comment by daeks** [comment]
well, it is a very small injection, as you already got access to write complete code files on the server (the  porpose of an IDE). Personally I would remove the complete system part and let the zip extentsion do the job (as it is already included, but not as primary option)

**Comment by evertton** [comment]
Please check this new PR. :smile:
